### ISSUE TITLE
`skipNonExistentBranches_toggleListingCommits_slideOutRoot` UI test spuriously fails: `java.net.SocketTimeoutException: timeout`

### ISSUE THREAD

**Issue by azdrojowa123** [VirtusLab/git-machete-intellij-plugin#830]
<details>
<summary>Logs</summary>
<pre>
com.virtuslab.gitmachete.uitest.UITestSuite > skipNonExistentBranches_toggleListingCommits_slideOutRoot FAILED
    java.net.SocketTimeoutException: timeout
        at okio.Okio$4.newTimeoutException(Okio.java:232)
        at okio.AsyncTimeout.exit(AsyncTimeout.java:286)
        at okio.AsyncTimeout$2.read(AsyncTimeout.java:241)
        at okio.RealBufferedSource.indexOf(RealBufferedSource.java:358)
        at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:230)
        at okhttp3.internal.http1.Http1ExchangeCodec.readHeaderLine(Http1ExchangeCodec.java:242)
        at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:213)
        at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.java:115)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:94)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229)
        at okhttp3.RealCall.execute(RealCall.java:81)
        at retrofit2.OkHttpCall.execute(OkHttpCall.java:204)
        at com.intellij.remoterobot.client.IdeRobotClient.retrieve(IdeRobotClient.kt:75)
        at com.intellij.remoterobot.JavaScriptApi$DefaultImpls.callJs(JavaScriptApi.kt:78)
        at com.intellij.remoterobot.RemoteRobot.callJs(RemoteRobot.kt:32)
        at com.virtuslab.gitmachete.uitest.RunningIntelliJFixtureExtension$RunningIntelliJFixtureOps.com$virtuslab$gitmachete$uitest$RunningIntelliJFixtureExtension$RunningIntelliJFixtureOps$$callJs(RunningIntelliJFixtureExtension.scala:39)
        at com.virtuslab.gitmachete.uitest.RunningIntelliJFixtureExtension$RunningIntelliJFixtureOps$project$.refreshModelAndGetRowCount(RunningIntelliJFixtureExtension.scala:135)
        at com.virtuslab.gitmachete.uitest.UITestSuite.skipNonExistentBranches_toggleListingCommits_slideOutRoot(UITestSuite.scala:127)

        Caused by:
        java.net.SocketException: Socket closed
            at java.base/java.net.SocketInputStream.read(SocketInputStream.java:183)
            at java.base/java.net.SocketInputStream.read(SocketInputStream.java:140)
            at okio.Okio$2.read(Okio.java:140)
            at okio.AsyncTimeout$2.read(AsyncTimeout.java:237)
            ... 27 more
</pre>
</details>

**Comment by PawelLipski** [comment]
Errors like that are notoriously difficult to debug... esp. that they might never happen locally.
B/c the only moment when you can actually peek into the error is when it actually happens (which is completely out of our control 😅)... let's put a mechanism that would actually **dump the state of IDE process** in case of an error.

I'd suggest running `jstack <PID>` against the IDE process (TBD how to extract its PID from ide-probe) once `SocketTimeoutException` occurs, and print `jstack`'s output directly to terminal and/or save it in CircleCI artifacts. This way we'll have some insight on what happened in the **IDE process** while the timeout occurred in the **testing process** (they're two separate processes!). My suspicion is that IDE process gets stuck on some **modal dialog**, which for some reason only activates once every few dozen test runs.

**Comment by PawelLipski** [comment]
Another such build: https://app.circleci.com/pipelines/github/VirtusLab/git-machete-intellij-plugin/4611/workflows/d5671ec8-a18e-4c1b-85fc-fc251b769b10/jobs/4839 ; again `skipNonExistentBranches_toggleListingCommits_slideOutRoot`, but different version (`2021.2.4`)

**Comment by PawelLipski** [comment]
Another step, if the above thread dump doesn't help, would be to **record the video** of the running IDE in the CI.
See https://www.google.com/search?q=xvfb-run+record+video (disclaimer: I haven't tested such `ffmpeg` usage; this might not work on Mac, a Linux VM might be needed for local test)

**Comment by PawelLipski** [comment]
Un-assigning from @azdrojowa123 as now all we can do is to wait for the error to happen again and read the thread dump 🧵

**Comment by PawelLipski** [comment]
We didn't need to wait long 😅 Here we go: https://app.circleci.com/pipelines/github/VirtusLab/git-machete-intellij-plugin/4932/workflows/8c803e9a-d69a-4571-ac90-4ca0b3a73f9e/jobs/5179/artifacts

The culprit seems to be a modal dialog (`DeleteBranchOnSlideOutSuggestionDialog`) indeed:

```
"AWT-EventQueue-0" #18 prio=6 os_prio=0 cpu=14524.41ms elapsed=257.32s tid=0x00007fb4400ed000 nid=0x530f waiting on condition  [0x00007fb43677a000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11.0.14.1/Native Method)
	- parking to wait for  <0x00000000c1d49120> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@11.0.14.1/LockSupport.java:194)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@11.0.14.1/AbstractQueuedSynchronizer.java:2081)
	at java.awt.EventQueue.getNextEvent(java.desktop@11.0.14.1/EventQueue.java:572)
	at com.intellij.ide.IdeEventQueue.lambda$getNextEvent$8(IdeEventQueue.java:598)
	at com.intellij.ide.IdeEventQueue$$Lambda$361/0x00000001005cbc40.compute(Unknown Source)
	at com.intellij.openapi.application.impl.ApplicationImpl.runUnlockingIntendedWrite(ApplicationImpl.java:831)
	at com.intellij.ide.IdeEventQueue.getNextEvent(IdeEventQueue.java:598)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(java.desktop@11.0.14.1/EventDispatchThread.java:190)
	at java.awt.EventDispatchThread.pumpEventsForFilter(java.desktop@11.0.14.1/EventDispatchThread.java:124)
	at java.awt.EventDispatchThread.pumpEventsForFilter(java.desktop@11.0.14.1/EventDispatchThread.java:117)
	at java.awt.WaitDispatchSupport$2.run(java.desktop@11.0.14.1/WaitDispatchSupport.java:190)
	at java.awt.WaitDispatchSupport$4.run(java.desktop@11.0.14.1/WaitDispatchSupport.java:235)
	at java.awt.WaitDispatchSupport$4.run(java.desktop@11.0.14.1/WaitDispatchSupport.java:233)
	at java.security.AccessController.doPrivileged(java.base@11.0.14.1/Native Method)
	at java.awt.WaitDispatchSupport.enter(java.desktop@11.0.14.1/WaitDispatchSupport.java:233)
	at java.awt.Dialog.show(java.desktop@11.0.14.1/Dialog.java:1070)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl$MyDialog.show(DialogWrapperPeerImpl.java:701)
	at com.intellij.openapi.ui.impl.DialogWrapperPeerImpl.show(DialogWrapperPeerImpl.java:437)
	at com.intellij.openapi.ui.DialogWrapper.doShow(DialogWrapper.java:1671)
	at com.intellij.openapi.ui.DialogWrapper.show(DialogWrapper.java:1629)
	at com.intellij.openapi.ui.DialogWrapper.showAndGet(DialogWrapper.java:1643)
	at com.virtuslab.gitmachete.frontend.actions.dialogs.DeleteBranchOnSlideOutSuggestionDialog.showAndGetSlideOutOptions(DeleteBranchOnSlideOutSuggestionDialog.kt:34)
	at com.virtuslab.gitmachete.frontend.actions.base.BaseSlideOutBranchAction.suggestBranchDeletion(BaseSlideOutBranchAction.java:140)
	at com.virtuslab.gitmachete.frontend.actions.base.BaseSlideOutBranchAction.lambda$deleteBranchIfRequired$3(BaseSlideOutBranchAction.java:126)
	at com.virtuslab.gitmachete.frontend.actions.base.BaseSlideOutBranchAction$$Lambda$4834/0x00000001027eb440.run(Unknown Source)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:214)
	at com.intellij.openapi.application.TransactionGuardImpl.access$200(TransactionGuardImpl.java:21)
	at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:196)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$invokeLater$4(ApplicationImpl.java:348)
	at com.intellij.openapi.application.impl.ApplicationImpl$$Lambda$337/0x0000000100586c40.run(Unknown Source)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:82)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:131)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:47)
	at com.intellij.openapi.application.impl.FlushQueue$FlushNow.run(FlushQueue.java:187)
	at java.awt.event.InvocationEvent.dispatch(java.desktop@11.0.14.1/InvocationEvent.java:313)
	at java.awt.EventQueue.dispatchEventImpl(java.desktop@11.0.14.1/EventQueue.java:776)
	at java.awt.EventQueue$4.run(java.desktop@11.0.14.1/EventQueue.java:727)
	at java.awt.EventQueue$4.run(java.desktop@11.0.14.1/EventQueue.java:721)
	at java.security.AccessController.doPrivileged(java.base@11.0.14.1/Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(java.base@11.0.14.1/ProtectionDomain.java:85)
	at java.awt.EventQueue.dispatchEvent(java.desktop@11.0.14.1/EventQueue.java:746)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:891)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:760)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:447)
	at com.intellij.ide.IdeEventQueue$$Lambda$363/0x00000001005e8c40.compute(Unknown Source)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:818)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:446)
	at com.intellij.ide.IdeEventQueue$$Lambda$362/0x00000001005cb440.run(Unknown Source)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:492)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(java.desktop@11.0.14.1/EventDispatchThread.java:203)
	at java.awt.EventDispatchThread.pumpEventsForFilter(java.desktop@11.0.14.1/EventDispatchThread.java:124)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(java.desktop@11.0.14.1/EventDispatchThread.java:113)
	at java.awt.EventDispatchThread.pumpEvents(java.desktop@11.0.14.1/EventDispatchThread.java:109)
	at java.awt.EventDispatchThread.pumpEvents(java.desktop@11.0.14.1/EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.run(java.desktop@11.0.14.1/EventDispatchThread.java:90)
```

**Comment by PawelLipski** [comment]
My suggestion now would be to add extra logging (`System.out.println`) in Rhino code, esp. around `findAndClickButton`... to make sure that e.g. the dialog indeed disappears after `robot.click(button)`. I mean... it apparently doesn't disappear for some reason :/

Probably still **too early** to go hard-code and record the entire session with `ffmpeg` (as it's probably non-trivial to configure).

**Comment by PawelLipski** [comment]
Actually! maybe even better than `System.out.println`s, instead of recording a **video**, better to just capture a **screenshot** from Xvfb, this should be rather easy &mdash; that's what worked on my Fedora:

```
xwd -display :99 -root -silent | convert xwd:- png:/tmp/screenshot.png
```

Things to figure out:
1. add `xwd` and ImageMagick tools (`convert`) to the CI image
2. whether the display number will always be `:99` (see how ide-probe invokes `xvfb-run`, probably `-a` or `-d` option is used)

Reassigning to @azdrojowa123 as this solution is more within her range of interests, I suppose 🙃

**Comment by PawelLipski** [comment]
Looks that it **might've** been fixed by @mkondratek's PR #927.... but as all such spurious errors, let's not close it (yet) 😅 let's run a few hundred builds to verify it doesn't happen anymore
